### PR TITLE
Performance: add preallocation for some slice/map

### PR DESCRIPTION
**Pull request by peakle**
**What is this feature?**

Added some slice/map preallocations for cases where we know the final slice/map size. This optimization allows us to reduce the number of calls to the go allocator / OS api and thereby reduce the program execution time.

**Who is this feature for?**

It's perfomance improvement.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/notifier/channels/alertmanager.go
Comment (peakle) [inline pkg/services/ngalert/notifier/channels/alertmanager.go:46]: here name collides with "url" package and variable is local so i renamed it to "u", but if it important i can go back this change, it not related to PR in general

**Comment by joeblubaugh** [comment]
Thanks for sending this, and looking into performance improvements in general.

Could you write up a proper PR description (it doesn't need to have all fields set, but something that will be readable for posterity would be great).

**Comment by peakle** [comment]
> Thanks for sending this, and looking into performance improvements in general.
> 
> Could you write up a proper PR description (it doesn't need to have all fields set, but something that will be readable for posterity would be great).

rewrote description, does it look good?

**Comment by peakle** [comment]
@joeblubaugh any updates?

**Review by joeblubaugh — APPROVED** [review]
Thanks for the contribution! There are one or two places where we will allocate a zero-length slice when there are no arguments, whereas before the pointer would just be nil. I think that's very unlikely to happen as these functions are usually called with non-zero arguments.

**Comment by peakle** [comment]
@joeblubaugh do i need to rollback this places?

**Comment by joeblubaugh** [comment]
No. I just needed to wait for CI to finish before merging.

File: pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go
Comment (joeblubaugh) [inline pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go:86]: The Cloud Platform test for the azure monitor is consistently failing in CI. Could you please first merge main? If that doesn't fix the problem, revert the change in this file and we'll see what CI does.

**Review by joeblubaugh — CHANGES_REQUESTED** [review]

File: pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go
Comment (andresmgot) [reply pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go:86]: we are aware of this issue, it's not related to the changes in this PR. @aangelisc is investigating it atm, will let you know once that is fixed. Sorry for the inconveniences @peakle

Comment (peakle) [reply pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go:86]: @joeblubaugh done
@andresmgot ok

Comment (andresmgot) [reply pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go:86]: if you want to unblock this PR you can remove this change so those tests are not triggered. There are usually not that many azure resourge graph queries within a request so the performance improvement here won't be noticeable.

Comment (peakle) [reply pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go]: @andresmgot good, btw i suppose create another perfomance pr, so i can send it later)

**Review by joeblubaugh — APPROVED** [review]

**Comment by marefr** [comment]
@joeblubaugh I think milestone should be 9.4.0 since v9.3.x branch already branched off from main. If 9.3.0 is needed you need to backport this to v9.3.x.

**Pull request merged**
